pulsar/transaction_coordinator_client.go

```diff
@@ -20,23 +20,322 @@ package pulsar
+			*res.Response.NewTxnResponse.TxnidLeastBits}
+	}
+}
+
+type addPublishPartitionOp struct {
+	// Request
+	id         *TxnID
+	partitions []string
+
+	// Response
+	done chan any
+	err  error
+}
+
+func (t *transactionHandler) addPublishPartitionToTxn(op *addPublishPartitionOp) {
+	requestID := t.tc.client.rpcClient.NewRequestID()
+	cmdAddPartitions := &pb.CommandAddPartitionToTxn{
+		RequestId:      proto.Uint64(requestID),
+		TxnidMostBits:  proto.Uint64(op.id.MostSigBits),
+		TxnidLeastBits: proto.Uint64(op.id.LeastSigBits),
+		Partitions:     op.partitions,
+	}
+	res, err := t.tc.client.rpcClient.RequestOnCnx(t.getConn(), requestID,
+		pb.BaseCommand_ADD_PARTITION_TO_TXN, cmdAddPartitions)
+	if t.checkRetriableError(err, op) {
+		return
+	}
+	defer close(op.done)
+	if err != nil {
+		op.err = err
+	} else if res.Response.AddPartitionToTxnResponse.Error != nil {
+		op.err = getErrorFromServerError(res.Response.AddPartitionToTxnResponse.Error)
+	}
+}
+
+type addSubscriptionOp struct {
+	// Request
+	id           *TxnID
+	topic        string
+	subscription string
+
+	// Response
+	done chan any
+	err  error
+}
+
+func (t *transactionHandler) addSubscriptionToTxn(op *addSubscriptionOp) {
+	requestID := t.tc.client.rpcClient.NewRequestID()
+	sub := &pb.Subscription{
+		Topic:        &op.topic,
+		Subscription: &op.subscription,
+	}
+	cmdAddSubscription := &pb.CommandAddSubscriptionToTxn{
+		RequestId:      proto.Uint64(requestID),
+		TxnidMostBits:  proto.Uint64(op.id.MostSigBits),
+		TxnidLeastBits: proto.Uint64(op.id.LeastSigBits),
+		Subscription:   []*pb.Subscription{sub},
+	}
+	res, err := t.tc.client.rpcClient.RequestOnCnx(t.getConn(), requestID,
+		pb.BaseCommand_ADD_SUBSCRIPTION_TO_TXN, cmdAddSubscription)
+	if t.checkRetriableError(err, op) {
+		return
+	}
+	defer close(op.done)
+	if err != nil {
+		op.err = err
+	} else if res.Response.AddSubscriptionToTxnResponse.Error != nil {
+		op.err = getErrorFromServerError(res.Response.AddSubscriptionToTxnResponse.Error)
+	}
+}
+
+type endTxnOp struct {
+	// Request
+	id     *TxnID
+	action pb.TxnAction
+
+	// Response
+	done chan any
+	err  error
+}
+
+func (t *transactionHandler) endTxn(op *endTxnOp) {
+	requestID := t.tc.client.rpcClient.NewRequestID()
+	cmdEndTxn := &pb.CommandEndTxn{
+		RequestId:      proto.Uint64(requestID),
+		TxnAction:      &op.action,
+		TxnidMostBits:  proto.Uint64(op.id.MostSigBits),
+		TxnidLeastBits: proto.Uint64(op.id.LeastSigBits),
+	}
+	res, err := t.tc.client.rpcClient.RequestOnCnx(t.getConn(), requestID, pb.BaseCommand_END_TXN, cmdEndTxn)
+	if t.checkRetriableError(err, op) {
+		return
+	}
+	defer close(op.done)
+	if err != nil {
+		op.err = err
+	} else if res.Response.EndTxnResponse.Error != nil {
+		op.err = getErrorFromServerError(res.Response.EndTxnResponse.Error)
+	}
+}
+
+func (t *transactionHandler) close() {
+	if t.getState() != txnHandlerReady {
+		return
+	}
+	close(t.closeCh)
+	t.setState(txnHandlerClosed)
```

Comment:
```suggestion
	if t.state.CAS(txnHandlerReady, txnHandlerClosed) {
		close(t.closeCh)
	}
```

Use a thread safe implementation. (then you can remove `setState`)